cnd/src/seed.rs

```diff
@@ -40,6 +47,141 @@ impl Seed {
         rand.try_fill(&mut arr[..])?;
         Ok(Seed(arr))
     }
+
+    /// Construct a seed from base64 data read in from pem file.
+    pub fn from_file<D: AsRef<OsStr>>(seed_file: D) -> Result<Seed, Error> {
+        let file = Path::new(&seed_file);
+        log::info!(
+            "Found secret seed file, reading from {}",
+            PrintablePath(&file.to_path_buf())
+        );
+
+        let contents = fs::read_to_string(file)?;
+        let pem = pem::parse(contents)?;
+
+        Seed::from_pem(pem)
+    }
+
+    pub fn from_pem(pem: pem::Pem) -> Result<Seed, Error> {
+        if pem.contents.len() != SEED_LENGTH {
+            return Err(Error::IncorrectLength(pem.contents.len()));
+        }
+
+        let mut array = [0; SEED_LENGTH];
+        for (i, b) in pem.contents.iter().enumerate() {
+            array[i] = *b;
+        }
+
+        Ok(Seed::from(array))
+    }
+
+    /// Read the seed from the default location if it exists, otherwise
+    /// generate a random seed and write it to the default location.
+    pub fn from_default_file_or_generate<R: Rng>(rand: R) -> Result<Seed, Error> {
+        let path = default_seed_path()?;
+
+        if path.exists() {
+            return Self::from_file(&path);
+        }
+
+        let random_seed = Seed::new_random(rand)?;
+        random_seed.write_to(path.clone())?;
+
+        log::info!(
+            "No seed file found, creating default at {}",
+            PrintablePath(&path)
+        );
+
+        Ok(random_seed)
+    }
+
+    fn write_to(&self, seed_file: PathBuf) -> Result<(), Error> {
+        ensure_directory_exists(seed_file.clone())?;
+        self._write_to(seed_file)?;
+        Ok(())
+    }
+
+    fn _write_to(&self, path: PathBuf) -> Result<(), Error> {
+        let pem = Pem {
+            tag: String::from("SECRET SEED"),
+            contents: self.0.to_vec(),
+        };
+
+        let pem_string = encode(&pem);
+
+        let mut file = File::create(path.clone())?;
+        file.write_all(pem_string.as_bytes())?;
+
+        Ok(())
+    }
+}
+
+fn ensure_directory_exists(file: PathBuf) -> Result<(), Error> {
```

Comment:
:question: Does this need to take an owned `PathBuf`?

Reply:
Yes, IIUC `PathBuf` is like `String` and parameters typically always consume it.  Discussed offline, posting here in case I'm wrong.